### PR TITLE
(Draft) TF / keras model spark UDF supports multi dimension input, force log model with signature

### PR DESCRIPTION
**Pull request by WeichenXu123**
## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

TF / keras model spark UDF supports multi dimension input
Force log model with signature
Make Keras pyfunc model only accepts numpy input.

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -221,6 +221,7 @@
 
 import numpy as np
 import pandas
+import pandas as pd
```
Comment (BenWilson2) [inline mlflow/pyfunc/__init__.py:224]: Should the non-aliased package import be removed?

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -359,6 +360,17 @@ def _load_model_env(path):
     return _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME).get(ENV, None)
 
 
+def _reshape_column_values(name, pd_series, shape, dtype):
+    flattened_numpy_arr = np.concatenate(pd_series.tolist())
+    reshaped_numpy_arr = flattened_numpy_arr.reshape(shape).astype(type)
```
Comment (BenWilson2) [inline mlflow/pyfunc/__init__.py:365]:
```suggestion
    reshaped_numpy_arr = flattened_numpy_arr.reshape(shape).astype(dtype)
```

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -398,6 +410,45 @@ def predict(self, data: PyFuncInput) -> PyFuncOutput:
...
+
+        if input_schema.is_tensor_spec() and isinstance(data, pd.DataFrame):
+            """
+            For tensor spec input schema, the model only accepts numpy array input or
+            dict type input (dict only contains numpy type values and keys must match
+            the input schema field names).
+            So this part code is to convert the pandas dataframe into a numpy array or
+            a dict. Note it contains some necessary array reshaping because the pandas
+            dataframe only support 1-dimension array value, so here it will reshape the
+            input values as the required shape according to schema field tensor spec.
```
Comment (BenWilson2) [inline mlflow/pyfunc/__init__.py:422]:
```suggestion
            This portion of the code is to convert the pandas dataframe into either a numpy array or
            a dict. Note that array reshaping is required due to pandas
            dataframe supporting only a 1-dimension array values, so a reshaping of the
            input values is required to conform the field to the tensor spec.
```

File: mlflow/tensorflow/__init__.py [hunk 4]
```diff
@@ -294,6 +282,33 @@ def save_model(
...
+        )
+    if len(signature.inputs.inputs) == 0:
+        raise MlflowException(
+            "The signature inputs must contain at least one field.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    for input in signature.inputs.inputs():
+        if not isinstance(input, TensorSpec):
+            raise MlflowException(
+                "All fileds in signature inputs must be tensor type.",
```
Comment (BenWilson2) [inline mlflow/tensorflow/__init__.py:298]:
```suggestion
                "All fields in signature inputs must be of tensor type.",
```

File: mlflow/tensorflow/__init__.py [hunk 5]
```diff
@@ -294,6 +282,33 @@ def save_model(
...
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        if input.name is None:
+            raise MlflowException(
+                "All fileds in signature inputs must have a name.",
```
Comment (BenWilson2) [inline mlflow/tensorflow/__init__.py:303]:
```suggestion
                "All fields in signature inputs must have a name.",
```

File: mlflow/tensorflow/__init__.py [hunk 6]
```diff
@@ -294,6 +282,33 @@ def save_model(
...
+                error_code = INVALID_PARAMETER_VALUE,
+            )
+        if input.shape[0] != -1:
+            raise MlflowException(
+                "All fileds in signature inputs must have shape that the first dimension "
+                "is a variable dimension."
```
Comment (BenWilson2) [inline mlflow/tensorflow/__init__.py:309]:
```suggestion
                "All fields in signature inputs must have a shape in which the first dimension "
                "is a variable dimension."
```

File: mlflow/tensorflow/__init__.py [hunk 7]
```diff
@@ -294,6 +282,33 @@ def save_model(
     import tensorflow
     from tensorflow.keras.models import Model as KerasModel
 
+    if signature is None:
+        raise MlflowException(
```
Comment (BenWilson2) [inline mlflow/tensorflow/__init__.py:286]: Should we add a warning in the docs about this breaking change? A note that specifies something along the lines of "MLflow 2.0 requires that TensorFlow and tf.Keras models are logged with a signature. A signature can be created for logging as shown below:" <brief toy example of generating a signature for a TF model>

**Review by BenWilson2 — COMMENTED** [review]
Some minor comments. With this level of a breaking change to the TF logging API, perhaps it is worthwhile including this in the 2.0 release :)

**Comment by harupy** [comment]
Let's change the base branch to master.

**Comment by harupy** [comment]
Can we add tests?

**Comment by WeichenXu123** [comment]
create a new PR against master instead https://github.com/mlflow/mlflow/pull/7291

**Pull request closed without merging**